File: frontend/src/types/weather.ts
export interface WeatherObservation {
  id: number
  farm_id: number
  temperature?: number
  humidity?: number
  rainfall?: number
  rainfall_probability?: number
  wind_speed?: number
  pressure?: number
  recorded_at: string
  source: string
}

export interface WeatherForecast {
  id: number
  farm_id: number
  forecast_time: string
  temperature?: number
  humidity?: number
  rainfall_probability?: number
  rainfall_amount?: number
  wind_speed?: number
  source: string
}

File: backend/app/models/weather.py
from datetime import datetime
from sqlalchemy import Column, Integer, Float, String, DateTime, ForeignKey
from sqlalchemy.sql import func

from app.database.session import Base


class WeatherObservation(Base):
    __tablename__ = "weather_observations"

    id = Column(Integer, primary_key=True, index=True)
    farm_id = Column(Integer, ForeignKey("farms.id"), nullable=False)
    temperature = Column(Float, nullable=True)
    humidity = Column(Float, nullable=True)
    rainfall = Column(Float, nullable=True)
    wind_speed = Column(Float, nullable=True)
    pressure = Column(Float, nullable=True)
    recorded_at = Column(DateTime(timezone=True), server_default=func.now())
    source = Column(String, nullable=False)


class WeatherForecast(Base):
    __tablename__ = "weather_forecasts"

    id = Column(Integer, primary_key=True, index=True)
    farm_id = Column(Integer, ForeignKey("farms.id"), nullable=False)
    forecast_time = Column(DateTime(timezone=True), nullable=False)
    temperature = Column(Float, nullable=True)
    humidity = Column(Float, nullable=True)
    rainfall_probability = Column(Float, nullable=True)
    rainfall_amount = Column(Float, nullable=True)
    wind_speed = Column(Float, nullable=True)
    source = Column(String, nullable=False)
